File: api/merge.js
const { IncomingForm } = require("formidable");
const XLSX = require("xlsx");
const fs = require("fs");

module.exports = async function handler(req, res) {
  try {
    res.setHeader("Access-Control-Allow-Origin", "*");
    res.setHeader("Access-Control-Allow-Methods", "POST, OPTIONS");
    res.setHeader("Access-Control-Allow-Headers", "Content-Type");

    if (req.method === "OPTIONS") {
      return res.status(200).end();
    }

    if (req.method !== "POST") {
      return res.status(405).send("Method not allowed");
    }

    const form = new IncomingForm({ multiples: false, keepExtensions: true });

    form.parse(req, async (err, fields, files) => {
      try {
        if (err) throw err;

        const workbookEntry = Array.isArray(files.workbook)
          ? files.workbook[0]
          : files.workbook;

        if (!workbookEntry) {
          return res.status(400).send("Missing workbook");
        }

        const workbookPath =
          workbookEntry.filepath ||
          workbookEntry.path ||
          workbookEntry.tempFilePath ||
          null;

        if (!workbookPath) {
          return res.status(500).send(
            "Workbook upload received, but no readable file path was found."
          );
        }

        const rawData = Array.isArray(fields.data) ? fields.data[0] : fields.data;
        if (!rawData) {
          return res.status(400).send("Missing data");
        }

        const rows = JSON.parse(rawData);

        const buffer = fs.readFileSync(workbookPath);
        const workbook = XLSX.read(buffer, { type: "buffer" });

        const sheet = workbook.Sheets["1_Data Entry"];
        if (!sheet) {
          return res.status(400).send('Sheet "1_Data Entry" not found');
        }

        rows.forEach((row, i) => {
          const r = 6 + i;

          if (row.ticker != null && row.ticker !== "") sheet["E" + r] = { v: row.ticker };
          if (row.buyDate != null && row.buyDate !== "") sheet["G" + r] = { v: row.buyDate };
          if (row.sharesBought != null && row.sharesBought !== "") sheet["H" + r] = { v: Number(row.sharesBought) };
          if (row.costPerShare != null && row.costPerShare !== "") sheet["I" + r] = { v: Number(row.costPerShare) };
          if (row.sellDate != null && row.sellDate !== "") sheet["J" + r] = { v: row.sellDate };
          if (row.sharesSold != null && row.sharesSold !== "") sheet["K" + r] = { v: Number(row.sharesSold) };
          if (row.salePricePerShare != null && row.salePricePerShare !== "") sheet["L" + r] = { v: Number(row.salePricePerShare) };
          if (row.note != null && row.note !== "") sheet["N" + r] = { v: row.note };
        });

        const output = XLSX.write(workbook, {
          type: "buffer",
          bookType: "xlsx"
        });

        res.setHeader(
          "Content-Disposition",
          'attachment; filename="merged.xlsx"'
        );
        res.setHeader(
          "Content-Type",
          "application/vnd.openxmlformats-officedocument.spreadsheetml.sheet"
        );

        return res.status(200).send(output);
      } catch (innerErr) {
        console.error("INNER ERROR:", innerErr);
        return res.status(500).send(innerErr.message);
      }
    });
  } catch (outerErr) {
    console.error("OUTER ERROR:", outerErr);
    return res.status(500).send(outerErr.message);
  }
};

File: api/merge.py
from io import BytesIO
from fastapi import FastAPI, UploadFile, File, Form, Response, HTTPException
from openpyxl import load_workbook
from datetime import datetime
import json

app = FastAPI()

SHEET_NAME = "1_Data Entry"
START_ROW = 6

INPUT_COLUMNS = {
    "ticker": "E",
    "buyDate": "G",
    "sharesBought": "H",
    "costPerShare": "I",
    "sellDate": "J",
    "sharesSold": "K",
    "salePricePerShare": "L",
    "note": "N",
}

def clear_cell(ws, cell_ref: str):
    ws[cell_ref].value = None

def parse_date(value):
    if not value:
        return None
    try:
        return datetime.strptime(value, "%Y-%m-%d")
    except:
        return None

@app.get("/")
def root():
    return {"status": "ok"}

@app.post("/api/merge")
async def merge_post(
    workbook: UploadFile = File(...),
    data: str = Form(...)
):
    try:
        rows = json.loads(data)
        if not isinstance(rows, list):
            raise HTTPException(status_code=400, detail="Data must be a JSON array")

        content = await workbook.read()
        wb = load_workbook(BytesIO(content))

        if SHEET_NAME not in wb.sheetnames:
            raise HTTPException(status_code=400, detail=f'Sheet "{SHEET_NAME}" not found')

        ws = wb[SHEET_NAME]

        # Force Excel FULL recalculation
        wb.calculation_properties.fullCalcOnLoad = True

        # Clear input cells
        for row_num in range(6, 506):
            for col in INPUT_COLUMNS.values():
                clear_cell(ws, f"{col}{row_num}")

        # Write rows
        for i, row in enumerate(rows, start=START_ROW):
            if i > 505:
                raise HTTPException(status_code=400, detail="Too many rows")

            for field, col in INPUT_COLUMNS.items():
                value = row.get(field)
                if value in (None, ""):
                    continue

                cell = ws[f"{col}{i}"]

                # NUMBERS
                if field in ("sharesBought", "costPerShare", "sharesSold", "salePricePerShare"):
                    cell.value = float(value)

                # DATES (CRITICAL FIX)
                elif field in ("buyDate", "sellDate"):
                    dt = parse_date(value)
                    if dt:
                        cell.value = dt
                        cell.number_format = "m/d/yyyy"

                # TEXT
                else:
                    cell.value = value

        output = BytesIO()
        wb.save(output)
        output.seek(0)

        return Response(
            content=output.getvalue(),
            media_type="application/vnd.openxmlformats-officedocument.spreadsheetml.sheet",
            headers={
                "Content-Disposition": 'attachment; filename="merged.xlsx"',
                "Access-Control-Allow-Origin": "*",
            },
        )

    except HTTPException:
        raise
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))
